### ISSUE TITLE
Some documents claim that NPVs maintain local copy of the ledger, even though the current implementation does not.

### ISSUE THREAD

**Issue by sachikoy** [hyperledger-archives/fabric#2097]
## Description

Some documents on GitHub says that a non-validating peer maintains local copy of the ledger. However, by observing the behavior of NVPs, it seems they don’t maintain ledger data.
If there is no plan to support local copies of the ledger on NVP, documents should updated to avoid misunderstanding.
## Describe How to Reproduce

Similar descriptions can be found below:
- https://github.com/hyperledger/fabric/blob/03c751fcc72b3a9c89985b1c4c290233a6871995/docs/glossary.md
- https://github.com/hyperledger/fabric/blob/e8aeb4973670f916ae49254a7a28809ec8054612/docs/FAQ/usage_FAQ.md
- https://github.com/hyperledger/fabric/blob/960db9f57a5e19f1f097741c807c01cdbe86ada6/docs/protocol-spec.md#4-security-1

**Comment by polyrabbit** [comment]
+1

On start, I only see peer starts `peer reconnect service` and `peer discovery service` in the code. Would someone please point to me where in the code that a peer syncs the ledger? Thanks!

**Comment by joshhus** [comment]
@binhn is it safe to remove these references to a shared ledger copy on NVPs? @ibmmark Mark can you flag this one with a Documentation label please.

**Comment by polyrabbit** [comment]
Hi, could someone please tell me why NVP node does not replicate ledger from VP node, since that will be a great performance improvement when dealing with query?

**Comment by sachikoy** [comment]
@polyrabbit My guess is because it is difficult to keep NVPs really up to date without being part of PBFT. Validating Peers synchronise the ledger by using the PBFT protocol. This include not only consensus per each set of transactions but also the state transfer to recover from a lagged or broken ledger. It seems we'll need to replicate a similar mechanism between VP and NVP, which could introduce too much complexity.

**Comment by polyrabbit** [comment]
I have made my modification to enable NVP replicating ledger from VP, just a trial to get familiar with fabric code.

Should anyone be interested please see my [commits](https://github.com/polyrabbit/fabric/commits/master?author=polyrabbit)

**Comment by christo4ferris** [comment]
@polyrabbit a PR would make it easier to review your changes. @srderson or @binhn would one of you please weigh-in on this thread? thanks

**Comment by polyrabbit** [comment]
Hi @christo4ferris, thanks for your advice, I'm just wondering what's the point of NVP having the ledger, since NVPs delegate all the transactions(including query) to VPs, and NVP doesn't know which VP is the correct one to replicate from?

**Comment by srderson** [comment]
The original proposed purpose of the NVP, to my knowledge, was to keep a local copy of the ledger and support query only transactions. The goal was to reduce load on the VP as @polyrabbit speculated. Each NVP is designed to be connected to a single VP which it must trust.

If the NVP is not maintaining a local copy of the ledger, it sounds like a bug to me. @binhn @jeffgarratt ?

**Comment by binhn** [comment]
@srderson yes, NVP keeps up the ledger via block notification.  Note that NVP will go away in the [next iteration](https://github.com/hyperledger/fabric/wiki/Next-Consensus-Architecture-Proposal). The notion of NVP is replaced by the combination of Submitter, Endorsers, and Committer.
